### ISSUE TITLE
aws_cognito_user_pool_domain_invalid_domain fails

### ISSUE THREAD

**Issue by pguinardco** [terraform-linters/tflint#933]
When trying to use the following domain we fail tflint validation (yet it's a valid domain in terraform and AWS cognito).

```
resource "aws_cognito_user_pool_domain" "domain" {
  domain          = "ab-cdef.sso.example.com"
  certificate_arn = var.domain_certificate_arn
  user_pool_id    = aws_cognito_user_pool.pool.id
}
```

```
Error: "ab-cdef.sso.example.com" does not match valid pattern ^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$ (aws_cognito_user_pool_domain_invalid_domain)
  on /builds/sofiinc/devops/sofi-terraform/cognito/at-work/user_pool.at-work.tf line 77
Please, fix the errors printed above
```

```
diff --git a/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain.go b/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain.go
index 9cd2688b..243e3381 100644
--- a/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain.go
+++ b/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain.go
@@ -27,7 +27,7 @@ func NewAwsCognitoUserPoolDomainInvalidDomainRule() *AwsCognitoUserPoolDomainInv
                attributeName: "domain",
                max:           63,
                min:           1,
-               pattern:       regexp.MustCompile(`^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$`),
+               pattern:       regexp.MustCompile(`^[a-z0-9](?:[a-z0-9\-\.]{0,61}[a-z0-9])?$`),
        }
 }

@@ -77,7 +77,7 @@ func (r *AwsCognitoUserPoolDomainInvalidDomainRule) Check(runner *tflint.Runner)
                        if !r.pattern.MatchString(val) {
                                runner.EmitIssue(
                                        r,
-                                       fmt.Sprintf(`"%s" does not match valid pattern %s`, truncateLongMessage(val), `^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$`),
+                                       fmt.Sprintf(`"%s" does not match valid pattern %s`, truncateLongMessage(val), `^[a-z0-9](?:[a-z0-9\-\.]{0,61}[a-z0-9])?$`),
                                        attribute.Expr.Range(),
                                )
                        }
diff --git a/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain_test.go b/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain_test.go
index aa0f5f34..9ff7b4ac 100644
--- a/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain_test.go
+++ b/rules/awsrules/models/aws_cognito_user_pool_domain_invalid_domain_test.go
@@ -23,7 +23,7 @@ resource "aws_cognito_user_pool_domain" "foo" {
                        Expected: tflint.Issues{
                                {
                                        Rule:    NewAwsCognitoUserPoolDomainInvalidDomainRule(),
-                                       Message: `"auth example" does not match valid pattern ^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$`,
+                                       Message: `"auth example" does not match valid pattern ^[a-z0-9](?:[a-z0-9\-\.]{0,61}[a-z0-9])?$`,
                                },
                        },
                },
```

## Version

Error happens on archlinux with this:
TFLint v0.20.2
Terraform v0.13.3

Error does not happen on mac with this:
TFLint v0.20.2
Terraform v0.13.3

**Comment by pguinardco** [comment]
https://github.com/terraform-linters/tflint/pull/934

**Comment by bendrucker** [comment]
Your comment mentioning a difference on Linux versus macOS is pretty suspicious. Can't see how that could be possible.

In any case, as mentioned in #934, this rule along with most `aws_*` rules are generated from aws-sdk-go's JSON models.

https://github.com/aws/aws-sdk-go/blob/a3411680ac767bb37ff50c73e58c53e2426fd80a/models/apis/cognito-idp/2016-04-18/api-2.json#L755-L770
https://github.com/aws/aws-sdk-go/blob/a3411680ac767bb37ff50c73e58c53e2426fd80a/models/apis/cognito-idp/2016-04-18/api-2.json#L2989-L3000
https://github.com/aws/aws-sdk-go/blob/a3411680ac767bb37ff50c73e58c53e2426fd80a/models/apis/cognito-idp/2016-04-18/api-2.json#L3352-L3357

And you'll see that those models implement the documented requirements:

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateUserPoolDomain.html#API_CreateUserPoolDomain_RequestSyntax

The CloudFormation docs contain additional description:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpooldomain.html

> This string can include only lowercase letters, numbers, and hyphens. Don't use a hyphen for the first or last character. Use periods to separate subdomain names.

I don't know what to make of this contradiction. If you reach out to Amazon for clarification, feel free to update this thread. If the pattern is adjusted upstream, TFLint will pick it up when we update the SDK and generate the rules.

**Comment by wata727** [comment]
I tried to run `terraform apply` with the following version:

```
% terraform -v
Terraform v0.13.3
+ provider registry.terraform.io/hashicorp/aws v3.8.0
```

```hcl
resource "aws_cognito_user_pool_domain" "main" {
  domain       = "ab-cdef.sso.example.com"
  user_pool_id = aws_cognito_user_pool.example.id
}

resource "aws_cognito_user_pool" "example" {
  name = "example-pool"
}
```

```
% terraform apply
...
[TRUNCATE]
...

aws_cognito_user_pool_domain.main: Creating...

Error: Error creating Cognito User Pool Domain: InvalidParameterException: The domain name contains an invalid character. Domain names can only contain lower-case letters, numbers, and hyphens. Please enter a different name that follows this format: ^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$
```

Looking at this result, the validation is working correctly and the documentation looks incorrect. I believe that whether or not `terraform apply` results in an error is a good criterion.

**Comment by bendrucker** [comment]
Good confirmation. That exception means AWS returned a 400 for the request, so that would seem to confirm. @pguinardco If you have supporting details for your claim that this is valid, please do share.
